SQL: Introduce core support for multivalue fields

bpintea opened the pull request:
This introduces support for accessing multivalue fields, whose value can be retrieved with a newly added ARRAY function. Only value extraction is of concern for this PR.

`ARRAY` is only allowed in projections, will only accept a field name as parameter and can't be provided as another function's parameter.

Part of #68066.

bpintea committed:
Introduce core support for multivalue fields

This introduces support for multivalue fields, whose value can be
retrieved with a newly added ARRAY function. Only value extraction is of
concern here.

bpintea committed:
Style fix

Remove unused import

bpintea committed:
Fix spotless rules enforcement

Apply formatting of:
 ./gradlew :x-pack:plugin:sql:qa:server:spotlessApply

bpintea committed:
Address review comments

- renaming of methods and vars;
- multi-value handling delegated to renamed enum;
- more tests.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

bpintea committed:
Style fix

Remove unused import.

bpintea committed:
Have array types report same meta data as base

Plus add tests around the modifications.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

matriv commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypes.java:
```diff
@@ -151,6 +152,10 @@ public static boolean isSigned(DataType t) {
     }
 
+    public static boolean isArray(DataType t) {
+        return t instanceof ArrayDataType;
```
I'd rather have the isArray() as a class mathod (not static) where the basic data type returns false and array true, so we avoid `nstanceof`

bpintea replied:
That's also a possibility, indeed.
During an earlier analysis with Costin we discussed if non-array types should be aware of the array properties (my initial proposal was for DateType to have the `baseType` field and `isArray` method). The outcome however was to subclass (and thus `ArrayDataType` appeared).
A mixed approach would be to add the `isArray()` method in `DataType` always return `false` and in `ArrayDataType` always return `true`, but that wouldn't be a strict separation between the types anymore. But doable, still.

matriv replied:
`instanceof` has some performance impact, at least in the past jvms :), but it's also a personal preference to be honest, I prefer methods like the `isArray()`, they seem more elegant to me.

matriv commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -118,7 +118,7 @@ public void testMultiValueFieldWithMultiValueLeniencyDisabled() throws IOExcepti
             () -> doWithQuery(() -> esWithLeniency(false), "SELECT int, keyword FROM test", results -> {})
         );
-        assertTrue(expected.getMessage().contains("Arrays (returned by [int]) are not supported"));
+        assertTrue(expected.getMessage().contains("Cannot return multiple values for field [int]; call ARRAY(int) instead"));
```
how about `use` instead of `call`?

astefan replied:
+1 for `use`.

bpintea replied:
Replaced.

matriv commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -150,11 +150,12 @@ public void testMultiValueFields_InsideObjects_WithMultiValueLeniencyDisabled()
         );
-        assertTrue(expected.getMessage().contains("Arrays (returned by [object.intsubfield]) are not supported"));
+        assertTrue(expected.getMessage().contains("Cannot return multiple values for field [object.intsubfield]; " +
+            "call ARRAY(object.intsubfield) instead"));
```
same here

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypes.java:
```diff
@@ -83,6 +87,32 @@
+    public static final DataType KEYWORD_ARRAY      = of(KEYWORD);
+    public static final DataType TEXT_ARRAY         = of(TEXT);
+    // date
+    public static final DataType DATETIME_ARRAY     = of(DATETIME);
```
maybe add a comment that since DATE/TIME are SQL only types (not coming from index fields directly) currently arrays of those 2 types are not supported.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypes.java:
```diff
@@ -153,23 +185,46 @@
-            .filter(e -> e.esType() != null)
-            .collect(toUnmodifiableMap(DataType::esType, t -> t));
+        .filter(e -> e.esType() != null)
+        .filter(e -> e instanceof ArrayDataType == false)
```
those instance of can also be avoided if there is a method isArray() on the datatype (not static)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypes.java:
```diff
@@ -153,23 +185,46 @@
 
     static {
-        Map<String, DataType> sqlToEs = new HashMap<>(mapSize(45));
+        Map<String, DataType> sqlToEs = new HashMap<>(mapSize(45)); // TODO: is this magic number 1) needed & 2) up to date?
```
why not use TYPES.size() ? am I missing something?

bpintea replied:
I've adjusted it.

matriv reviewed (commented):
Looks good overall, Left a first round of comments.

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractorTests.java:
```diff
@@ -442,6 +470,204 @@ public void testFieldsWithSingleValueArrayAsSubfield_TwoNestedLists2() {
+        assertEquals(list, fe.extractFromSource(map));
+    }
+
+    // "a": [i1, i2, ..]
```
example doesn't show the difference from the prev test.

bpintea replied:
Extended comment.

matriv reviewed (commented):
- Please revert the whitespace changes on unrelated lines in the `.csv-spec` files
 - The PR is currently missing documentation, will it be added in upcoming PR?

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionTypeRegistry.java:
```diff
@@ -18,7 +18,8 @@
         GROUPING(GroupingFunction.class),
-        SCORE(Score.class);
+        SCORE(Score.class),
+        ARRAY(Array.class);
```
I think this list is meant to be ordered alphabetically.

matriv replied:
Not really, the functions within the groups are of course ordered, but the order of the groups it's up to us I think.

bpintea replied:
I've reordered them, just in case.

astefan reviewed (changes requested):
There are some aspects that might need a second look.
Also the importance of `field_multi_value_leniency` (and the documented limitation) I think needs a short discussion about its necessity.

bpintea committed:
Address review comments

Address review comments.

bpintea committed:
Undo stripping of trailing WS

Undo stripping of trailing WS in .csv-spec files.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -39,6 +42,41 @@
+            if (object instanceof List) { // is this a multi-value?
+                List<?> list = (List<?>) object;
+                if (list.size() > 1 && failIfMultiValue) {
+                    throw new QlIllegalArgumentException("Cannot return multiple values for field [{}]; call ARRAY({}) instead",
```
This is the equivalent message to current `Arrays (returned by [...]) are not supported` that doesn't hint about alternatives. But that might be b/c the `field_multi_value_leniency` appeared only later however.
Up to debate if we should advertise it, considering that some users can't control this parameter (like CLI; or drivers, though there it's always on).

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractorTests.java:
```diff
@@ -442,6 +470,204 @@ public void testFieldsWithSingleValueArrayAsSubfield_TwoNestedLists2() {
+                    try {
+                        source.value(x);
+                    } catch (IOException ignored) {
+                        fail();
```
.value() throws a checked exception so can't be called from `forEach`'s lambda.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -223,15 +272,15 @@ private static NumberType numberType(DataType dataType) {
 
         while (!queue.isEmpty()) {
-            Tuple<Integer, Map<String, Object>> tuple = queue.removeLast();
+            Tuple<Integer, Map<String, Object>> tuple = queue.removeFirst();
```
It's just to preserve the relative order of one doc's exploration.
With the example in [`testBasicQueryWithMultiValuesAndMultiPathAndMultiDoc`](https://github.com/bpintea/elasticsearch/blob/241dd43d1cec190d63affbea7ec1f06c44337941/x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java#L712) the returned value for `ARRAY(a.b.c.d)` returns now `[2, 3, 4, 5]` vs `[5, 3, 4, 2]` with the original tail dequeuing.

bpintea commented:
> Please revert the whitespace changes on unrelated lines in the .csv-spec files

I've reverted them, since the change was large for only two extra lines per file.
(Still, it'd be helpful if we'd remove this trailing white-space in one sweep sometimes.
Also, the gear icon menu on GH offers the option `Hide whitespace changes`, just as general FYI.) 

> The PR is currently missing documentation, will it be added in upcoming PR?

Yes, #68066 has a breakdown of the work.

matriv reviewed (approved):
LGTM. Thx @bpintea !

bpintea committed:
Explicitely verify for ARRAY in filtering

The WHERE clause needs a boolean condition, so ARRAY() used in WHERE
would already be caught by the check preventing it being used as an
argument to another function. However, the error message returned in
this case can be misleading.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

bpintea committed:
Correct data type ID calculation

... to work with the new DataTypes#isDateTime().

palesz reviewed (approved):
LGTM, only a few minor comments.

bpintea committed:
Adress review feedback

- remove unnecessary check.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/Array.java:
```diff
@@ -0,0 +1,73 @@
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() != 1) {
+            throw new IllegalArgumentException("expected [1] child but received [" + newChildren.size() + "]");
+        }
```
Thanks, I even had a todo for it. Fixed.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypes.java:
```diff
@@ -28,6 +29,7 @@
 import static java.util.Collections.unmodifiableMap;
 import static java.util.stream.Collectors.toUnmodifiableList;
 import static java.util.stream.Collectors.toUnmodifiableMap;
+import static org.elasticsearch.xpack.ql.type.ArrayDataType.of;
```
I'd keep the `of`, as it'd be a bit redundant otherwise , but maybe no longer import it static on conflict.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -226,15 +283,15 @@ private static NumberType numberType(DataType dataType) {
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
     public Object extractFromSource(Map<String, Object> map) {
```
I did have a thought about that too, though some checks would be done away with on refactoring. Now, if all the checks are actually needed -- not sure; but it's very likely that extraction from source will no longer be needed with the new fields API, so I'm waiting with further modifications till that change is opened.

palesz replied:
Not sure you'll lose any of the checks, but if this code is about to removed, leave it as is.

bpintea committed:
License headers fix

- update to new header.

astefan reviewed (approved):
LGTM

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/multivalue-core

bpintea committed:
Revert wrong merging

Pick updated code vs old-master one.
